core/common/src/main/java/alluxio/grpc/GrpcChannelKey.java

```diff
@@ -0,0 +1,316 @@
+    return this;
+  }
+
+  /**
+   * @return human readable name for the channel
+   */
+  public Optional<String> getChannelName() {
+    return mClientType;
+  }
+
+  /**
+   * Sets human readable name for the channel's client.
+   *
+   * @param clientType channel client type
+   * @return the modified {@link GrpcChannelKey}
+   */
+  public GrpcChannelKey setClientType(String clientType) {
+    mClientType = Optional.of(clientType);
+    return this;
+  }
+
+  /**
+   * @return keep alive time for the underlying channel
+   */
+  public Optional<Pair<Long, TimeUnit>> getKeepAliveTime() {
+    return mKeepAliveTime;
+  }
+
+  /**
+   * @param keepAliveTime keep alive time for the underlying channel
+   * @param timeUnit time unit for the keepAliveTime parameter
+   * @return the modified {@link GrpcChannelKey}
+   */
+  public GrpcChannelKey setKeepAliveTime(long keepAliveTime, TimeUnit timeUnit) {
+    mKeepAliveTime = Optional.of(new Pair<>(keepAliveTime, timeUnit));
+    return this;
+  }
+
+  /**
+   * @return keep alive timeout for the underlying channel
+   */
+  public Optional<Pair<Long, TimeUnit>> getKeepAliveTimeout() {
+    return mKeepAliveTimeout;
+  }
+
+  /**
+   * @param keepAliveTimeout keep alive timeout for the underlying channel
+   * @param timeUnit time unit for the keepAliveTimeout parameter
+   * @return the modified {@link GrpcChannelKey}
+   */
+  public GrpcChannelKey setKeepAliveTimeout(long keepAliveTimeout, TimeUnit timeUnit) {
+    mKeepAliveTimeout = Optional.of(new Pair<>(keepAliveTimeout, timeUnit));
+    return this;
+  }
+
+  /**
+   *
+   * @param strategy the pooling strategy
+   * @return the modified {@link GrpcChannelKey}
+   */
+  public GrpcChannelKey setPoolingStrategy(PoolingStrategy strategy) {
+    // TODO(feng): implement modularized pooling strategies
+    switch (strategy) {
+      case DEFAULT:
+        mPoolKey = 0L;
+        break;
+      case DISABLED:
+        mPoolKey = RANDOM.nextLong();
+        break;
+      default:
+        throw new IllegalArgumentException(
+            String.format("Invalid pooling strategy %s", strategy.name()));
+    }
+    return this;
+  }
+
+  @Override
+  public int hashCode() {
+    HashCodeBuilder hashCodebuilder = new HashCodeBuilder();
+    for (Field field : this.getClass().getDeclaredFields()) {
+      if (field.isAnnotationPresent(IdentityField.class)) {
+        try {
+          hashCodebuilder.append(field.get(this));
+        } catch (IllegalAccessException e) {
+          throw new RuntimeException(
+              String.format("Failed to calculate hashcode for channel-key: %s", this), e);
+        }
+      }
+    }
+    return hashCodebuilder.toHashCode();
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    if (other instanceof GrpcChannelKey) {
+      GrpcChannelKey otherKey = (GrpcChannelKey) other;
+      boolean areEqual = true;
+      for (Field field : this.getClass().getDeclaredFields()) {
```

Comment:
will this type of reflection impact performance?

also, can we add unit tests for equals and hashcode? (do they already exist?)

Reply:
This will not be in hot path, only to acquire a physical channel. Since our channels are long lived I don't think we would notice any change in performance.

`GrpcManagedChannelPoolTest` does some testing on this area.